File: Linked Lists/Problem 4/problem4.py
class Node:
    def __init__(self, data: int):
        self.data: int = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head: Node = None
    
    def append(self, data: int) -> None:
        newNode: Node = Node(data)
        newNode.next = None

        if (self.head is None):
            self.head = newNode
            self.head.next = self.head
            return
        
        temp: Node = self.head
        
        while (temp is not None):
            if (temp.next == self.head):
                temp.next = newNode
                newNode.next = self.head
                return

            temp = temp.next
    
    def betterAppedn(self, data: list[int]):
        for nums in data:
            self.append(nums)
    
    def printList(self):
        temp: Node = self.head

        while (temp is not None):
            print(f"{temp.data}->", end="")

            temp = temp.next

            if (temp == self.head):
                break
            
        print("NULL\n")

def main() -> int:
    list: LinkedList = LinkedList()

    list.betterAppedn([1, 2, 3, 4, 5, 6])
    list.printList()
    
    return 0

if __name__ == "__main__":
    main()        

File: Linked Lists/Problem 4/problem4.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

void append(Node **head, int data) {
    Node *newNode = (Node *)malloc(sizeof(Node));

    newNode->data = data;
    newNode->next = NULL;

    if (*head == NULL) {
        *head = newNode;
        (*head)->next = *head;
        return;
    }

    for (Node *temp = *head; temp != NULL; temp = temp->next) {
        if (temp->next == *head) {
            temp->next = newNode;
            newNode->next = *head;
            return;
        }
    }
}

void betterAppend(Node **head, int *data, int length) {
    for (int i = 0; i < length; i++) {
        append(head, data[i]);
    }
}

void printArray(Node *head) {
    for (Node *temp = head; temp != NULL; temp = temp->next) {
        printf("%d->", temp->data);

        if (temp->next == head) {
            break;
        }
    }

    printf("NULL\n");
}

int main(int argc, char const *argv[]) {
    Node *head = NULL;

    int data[] = {1, 2, 3, 4, 5, 6};
    betterAppend(&head, &data[0], (sizeof(data) / sizeof(data[0])));
    printArray(head);

    return 0;
}
